storagenode/pieces/store.go

```diff
@@ -46,42 +47,104 @@ type Info struct {
+	GetExpired(ctx context.Context, expiredAt time.Time, limit int64) ([]ExpiredInfo, error)
+	// SetExpiration sets an expiration time for the given piece ID on the given satellite
+	SetExpiration(ctx context.Context, satellite storj.NodeID, pieceID storj.PieceID, expiresAt time.Time) error
+	// DeleteExpiration removes an expiration record for the given piece ID on the given satellite
+	DeleteExpiration(ctx context.Context, satellite storj.NodeID, pieceID storj.PieceID) (found bool, err error)
+	// DeleteFailed marks an expiration record as having experienced a failure in deleting the
+	// piece from the disk
+	DeleteFailed(ctx context.Context, satelliteID storj.NodeID, pieceID storj.PieceID, failedAt time.Time) error
+}
+
+// V0PieceInfoDB stores meta information about pieces stored with storage format V0 (where
+// metadata goes in the "pieceinfo" table in the storagenodedb). The actual pieces are stored
+// behind something providing the storage.Blobs interface.
+type V0PieceInfoDB interface {
 	// Get returns Info about a piece.
 	Get(ctx context.Context, satelliteID storj.NodeID, pieceID storj.PieceID) (*Info, error)
-	// GetPieceIDs gets pieceIDs using the satelliteID
-	GetPieceIDs(ctx context.Context, satelliteID storj.NodeID, createdBefore time.Time, limit, offset int) (pieceIDs []storj.PieceID, err error)
 	// Delete deletes Info about a piece.
 	Delete(ctx context.Context, satelliteID storj.NodeID, pieceID storj.PieceID) error
 	// DeleteFailed marks piece deletion from disk failed
 	DeleteFailed(ctx context.Context, satelliteID storj.NodeID, pieceID storj.PieceID, failedAt time.Time) error
-	// SpaceUsed returns the in memory value for disk space used by all pieces
-	SpaceUsed(ctx context.Context) (int64, error)
-	// CalculatedSpaceUsed calculates disk space used by all pieces
-	CalculatedSpaceUsed(ctx context.Context) (int64, error)
-	// SpaceUsedBySatellite calculates disk space used by all pieces by satellite
-	SpaceUsedBySatellite(ctx context.Context, satelliteID storj.NodeID) (int64, error)
-	// GetExpired gets orders that are expired and were created before some time
+	// GetExpired gets piece IDs stored with storage format V0 that expire or have expired
+	// before the given time
 	GetExpired(ctx context.Context, expiredAt time.Time, limit int64) ([]ExpiredInfo, error)
+	// ForAllV0PieceIDsOwnedBySatellite executes doForEach for each locally stored piece, stored
+	// with storage format V0 in the namespace of the given satellite. If doForEach returns a
+	// non-nil error, ForAllV0PieceIDsOwnedBySatellite will stop iterating and return the error
+	// immediately.
+	ForAllV0PieceIDsOwnedBySatellite(ctx context.Context, blobStore storage.Blobs, satellite storj.NodeID, doForEach func(StoredPieceAccess) error) error
+}
+
+// V0PieceInfoDBForTest is like V0PieceInfoDB, but adds on the Add() method so
+// that test environments with V0 piece data can be set up.
+type V0PieceInfoDBForTest interface {
```

Comment:
Can this interface definition simply occur in a test file? I usually prefer to not have test-specific things inside non-test files if possible.

Reply:
It could, but then it could not be used in the packages that need it (none of which are storagenode/pieces).

Certainly keeping test-specific things in test files is preferable, but when it needs to be done, I very much like having interfaces like this that make it very clear when something is dealing with the test-specific functionality, instead of just letting it be mixed in with the standard functionality.

Reply:
:+1: